File: dlib-sys/src/image_window.rs
use crate::{
    cv_image::{self, CvImage},
    matrix::Matrix,
    overlay_line::OverlayLine,
    rectangle::Rectangle,
};
use autocxx::prelude::*;
use cxx::CxxVector;
use std::pin::Pin;

pub struct ImageWindow {
    pub(crate) inner: Pin<Box<crate::ffi::wrapper::ImageWindow>>,
}

impl Default for ImageWindow {
    fn default() -> Self {
        Self {
            inner: crate::ffi::wrapper::ImageWindow::new().within_box(),
        }
    }
}

impl ImageWindow {
    pub fn set_matrix(&mut self, matrix: &Matrix) {
        self.inner.as_mut().set_matrix(&matrix.inner.as_ref())
    }

    pub fn set_cv_image(&mut self, cv_image: &CvImage) {
        self.inner.as_mut().set_cv_image(&cv_image.inner.as_ref())
    }

    pub fn add_rectangle_overlay(&mut self, rectangle: Rectangle) {
        self.inner.as_mut().add_rectangle_overlay(rectangle.inner);
    }

    pub fn add_line_overlay(&mut self, overlay_lines: UniquePtr<CxxVector<crate::ffi::wrapper::OverlayLine>>) {
        self.inner.as_mut().add_line_overlay(overlay_lines);
    }

    pub fn clear_overlay(&mut self) {
        self.inner.as_mut().clear_overlay();
    }

    pub fn is_closed(&self) -> bool {
        self.inner.is_closed()
    }
}

unsafe impl Send for ImageWindow {}

File: dlib-sys/src/wrapper/image_window.cpp
#pragma once

#include <algorithm>
#include <dlib/gui_widgets.h>
#include <dlib/gui_widgets/widgets.h>
#include <exception>
#include <iterator>
#include <memory>
#include <vector>

#include "cv_image.cpp"
#include "matrix.cpp"
#include "overlay_line.cpp"
#include "rectangle.cpp"

namespace wrapper {

struct ImageWindow {
  dlib::image_window inner;

  void set_matrix(const Matrix &matrix) { inner.set_image(matrix.inner); }

  void set_cv_image(const CvImage &cv_image) {
    inner.set_image(cv_image.inner);
  }

  void add_rectangle_overlay(Rectangle rectangle) {
    this->inner.add_overlay(rectangle.inner);
  }

  void add_line_overlay(std::vector<OverlayLine> overlay_lines) {
    std::vector<dlib::image_window::overlay_line> dlib_overlay_lines;

    std::transform(overlay_lines.begin(), overlay_lines.end(),
                   std::back_inserter(dlib_overlay_lines),
                   [](OverlayLine x) { return x.inner; });

    this->inner.add_overlay(dlib_overlay_lines);
  }

  void clear_overlay() { this->inner.clear_overlay(); }

  bool is_closed() const {
    try {
      return this->inner.is_closed();
    } catch (const std::exception &e) {
      return false;
    }
  }
};
} // namespace wrapper
